cmd/crypto_cloud.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package cmd
+
+import (
+	"encoding/base64"
+
+	"github.com/pulumi/pulumi/pkg/secrets"
+	"github.com/pulumi/pulumi/pkg/secrets/cloud"
+	"github.com/pulumi/pulumi/pkg/tokens"
+	"github.com/pulumi/pulumi/pkg/util/contract"
+	"github.com/pulumi/pulumi/pkg/workspace"
+)
+
+func newCloudSecretsManager(stackName tokens.QName, configFile, secretsProvider string) (secrets.Manager, error) {
+	contract.Assertf(stackName != "", "stackName %s", "!= \"\"")
+
+	if configFile == "" {
+		f, err := workspace.DetectProjectStackPath(stackName)
+		if err != nil {
+			return nil, err
+		}
+		configFile = f
+	}
+
+	info, err := workspace.LoadProjectStack(configFile)
+	if err != nil {
+		return nil, err
+	}
+	info.SecretsProvider = secretsProvider
+
+	var secretsManager *cloud.Manager
+	// TODO: We shouldn't be re-using the `EncryptionSalt` field here - this is really a `DataKey`.
+	// TBD how to better represent this in the ProjectStack data structure.
```

Comment:
@ellismg Curious what you think we should do here?  It's odd that we have two places we need to store this secretsprovider state (`Pulumi.stackname.yaml` and the state file), but they currently use completely separate storage.  Should we somehow align these two more closely?  Or make `SecretsProvider` in the `Pulumi.stackname.yaml` actually a struct so it can include state associated with the secrets provider?

More fundamentally, do we really need to store secretsprovider state in both places?

Reply:
> More fundamentally, do we really need to store secretsprovider state in both places?

I believe that we definitively need it in the state file, because StackReference exists and when you take a reference to another stack you need to be able to deserialize the deployment, which requires information about the secrets provider.  It would not be possible to read this information from the `Pulumi.<stack-name>.yaml` file, as we have no idea where it is (in the case of referencing a different project it may not even be present).

It may be possible to remove the configuration from `Pulumi.<stack-name>.yaml` and just say we always have to fetch the deployment in order to understand secrets provider to use, but this would mean we'd have to do the logical equivalent of `pulumi stack export` when running a subset of `pulumi config` commands, and I'm not sure how we want to do that.

>  make SecretsProvider in the Pulumi.stackname.yaml actually a struct so it can include state associated with the secrets provider?

Maybe this is the way to go.  If we changed the shape of State from `json.RawMessage` to just `map[string]interface{}` the we would be able to store the same content in both the YAML and JSON files?

Reply:
I've gone ahead and used `encryptedkey` as the name of this field in both contexts.  For now I have not done any larger refactoring to reduce duplication.  We believe it is possible to do that in the future if we think it makes sense.